### PR TITLE
LOOKDEVX-470 : Look  and LookGroup management 

### PR DESCRIPTION
**Pull request by bernardkwok**
## Updates
Add in look / lookgroup management utilities
- Add in ability to append a look to another look
- Add ability to merge all looks together in a lookup
- Add ability to merge a lookup into another lookup with the option to skip duplicate looks.
Add string array to string merge utility.

## Testing
1. Starting configuration
```
<?xml version="1.0"?>
<materialx version="1.38">
  <lookgroup name="lookgroup1" looks="look1,look2,look3,look4,look5" active="look1" />
  <look name="look1">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look2">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look3">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look4">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look5">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <!-- Copy test -->
  <lookgroup name="lookgroup1_copy" looks="look1,look2,look3,look4,look5" active="look1" />
</materialx>
```
2. Results after merging looks in "lookgroup" to produce new look "look6"
```
<?xml version="1.0"?>
<materialx version="1.38">
  <lookgroup name="lookgroup1" looks="look6" active="look6" />
  <look name="look1">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look2">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look3">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look4">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look5">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <lookgroup name="lookgroup1_copy" looks="look1,look2,look3,look4,look5" active="look1" />
  <look name="look6">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
    <materialassign name="look2_matA" material="materialA" />
    <materialassign name="look2_matB" material="materialB" />
    <materialassign name="look2_matC" material="materialC" />
    <materialassign name="look3_matA" material="materialA" />
    <materialassign name="look3_matB" material="materialB" />
    <materialassign name="look3_matC" material="materialC" />
    <materialassign name="look4_matA" material="materialA" />
    <materialassign name="look4_matB" material="materialB" />
    <materialassign name="look4_matC" material="materialC" />
    <materialassign name="look5_matA" material="materialA" />
    <materialassign name="look5_matB" material="materialB" />
    <materialassign name="look5_matC" material="materialC" />
  </look>
</materialx>
```
3. Results after merging new lookgroup ("lookgroup2") with ("lookgroup1_copy"). 
```
<?xml version="1.0"?>
<materialx version="1.38">
  <look name="look1">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look2">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look3">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look4">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look5">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <lookgroup name="lookgroup1_copy" looks="look1,look2,look3,look4,look5" active="look1" />
  <lookgroup name="lookgroup2" looks="lookA,look1,lookC,look3,lookE" active="lookA,look3,lookE" />
  <look name="lookA">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="lookC">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="lookE">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <lookgroup name="lookgroup1_copy_merged" looks="look1, look2, look3, look4, look5, lookA, lookC, lookE" active="look1, lookA, look3, lookE" />
</materialx>
```
3. Results after trying to insert after "look2" and to look which is not found (lookgroup1_copy_merged2, and lookgroup1_copy_merged3 respectively)
```
<?xml version="1.0"?>
<materialx version="1.38">
  <look name="look1">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look2">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look3">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look4">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="look5">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <lookgroup name="lookgroup1_copy" looks="look1,look2,look3,look4,look5" active="look1" />
  <lookgroup name="lookgroup2" looks="lookA,look1,lookC,look3,lookE" active="lookA,look3,lookE" />
  <look name="lookA">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="lookC">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <look name="lookE">
    <materialassign name="matA" material="materialA" />
    <materialassign name="matB" material="materialB" />
    <materialassign name="matC" material="materialC" />
  </look>
  <lookgroup name="lookgroup1_copy_merged" looks="look1, look2, look3, look4, look5, lookA, lookC, lookE" active="look1, lookA, look3, lookE" />
  <lookgroup name="lookgroup1_copy_merged2" looks="look1, look2, lookA, lookC, lookE, look3, look4, look5" active="look1, lookA, look3, lookE" />
  <lookgroup name="lookgroup1_copy_merged3" looks="look1, look2, look3, look4, look5, lookA, lookC, lookE" active="look1, lookA, look3, lookE" />
</materialx>
```

### CODE REVIEW

File: source/MaterialXCore/Look.h [hunk 1]
```diff
@@ -235,6 +239,11 @@ class Look : public Element
     }
 
     /// @}
+    /// @name Operators
+    /// @{
+
+    /// Append another look to this look
+    void append(const LookPtr& look);
```
Comment (bernardkwok) [inline source/MaterialXCore/Look.h:246]: Support method for merging looks together.

File: source/MaterialXCore/Look.h [hunk 2]
```diff
@@ -275,6 +284,16 @@ class LookGroup : public Element
         return getAttribute(ACTIVE_ATTRIBUTE);
     }
 
+    /// Return list of look Elements which are active
+    LookVec getActiveLooks() const;
```
Comment (bernardkwok) [inline source/MaterialXCore/Look.h:288]: Utility to hide trouble of parsing strings and finding looks.

File: source/MaterialXCore/Look.h [hunk 3]
```diff
@@ -275,6 +284,16 @@ class LookGroup : public Element
...
+
+    /// Append the contents of another lookgroup to this lookgroup. 
+    /// Optionally allow duplicate looks to be appended. Default is to disallow duplicates.
+    void append(const LookGroupPtr& lookgroup, bool skipDuplicateLooks = true);
```
Comment (bernardkwok) [inline source/MaterialXCore/Look.h:292]: For merging lookgroups together.

File: source/MaterialXCore/Look.h [hunk 4]
```diff
@@ -275,6 +284,16 @@ class LookGroup : public Element
...
+
+    /// Get a single combined look wihch contains the contents of all the looks in the lookgroup
+    LookPtr combineLooks();
```
Comment (bernardkwok) [inline source/MaterialXCore/Look.h:295]: For merging looks in a lookgroup.

File: source/MaterialXCore/Util.cpp [hunk 5]
```diff
@@ -91,6 +91,16 @@ StringVec splitString(const string& str, const string& sep)
     return split;
 }
 
+string mergeStringVec(const StringVec& stringVec, const string& sep)
```
Comment (bernardkwok) [inline source/MaterialXCore/Util.cpp:94]: Adding this utility for common reuse.

File: source/MaterialXTest/MaterialXCore/Look.cpp [hunk 6]
```diff
@@ -98,24 +101,90 @@ TEST_CASE("LookGroup", "[look]")
     std::vector<mx::LookGroupPtr> lookGroups = doc->getLookGroups();
     REQUIRE(lookGroups.size() == 1);
 
+    mx::StringMap mergeMap;
+
     const std::string looks = "look1,look2,look3,look4,look5";
-    mx::StringVec looksVec = mx::splitString(looks, ",");
+    mx::StringVec looksVec = mx::splitString(looks, mx::ARRAY_VALID_SEPARATORS);
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXCore/Look.cpp:107]: Note: We should be using ARRAY_VALID_SEPARATORS which allows for "," or ", " as the separator.

File: source/MaterialXCore/Look.cpp [hunk 7]
```diff
@@ -136,4 +158,121 @@ vector<VariantAssignPtr> MaterialAssign::getActiveVariantAssigns() const
...
+                appendLook = false;
+            }
+        }
+        if (appendLook)
+        {
+            destLookList.push_back(lookName);
+        }
+    }
+
+    // If look already exists in the active look list then append if not skipping duplicates
```
Comment (bernardkwok) [inline source/MaterialXCore/Look.cpp:224]: Note I didn't not add in any special casing here to handle the case where we may not want the default look to be here. i.e. it's always additive for now. We should talk about where this particular logic should belong separately.

File: source/MaterialXCore/Look.cpp [hunk 8]
```diff
@@ -136,4 +158,121 @@ vector<VariantAssignPtr> MaterialAssign::getActiveVariantAssigns() const
...
+    string looks = getActiveLook();
+    if (looks.empty())
+    {
+        looks = getLooks();
+    }
+    const StringVec& lookList = splitString(looks, ARRAY_VALID_SEPARATORS);
+    LookVec activeLooks;
+    if (!lookList.empty())
+    {
+        vector<LookPtr> lookElements = getDocument()->getLooks();
```
Comment (feldstj) [inline source/MaterialXCore/Look.cpp:176]: Should this also be const vector<LookPtr>& ?

File: source/MaterialXCore/Look.cpp [hunk 9]
```diff
@@ -136,4 +158,121 @@ vector<VariantAssignPtr> MaterialAssign::getActiveVariantAssigns() const
...
+        return;
+    }
+
+    // If look already exists in the look list then append if
+     // not skipping duplicates
+    const StringVec& sourceLookList = splitString(sourceLooks, ARRAY_VALID_SEPARATORS);
+    StringVec destLookList = splitString(getLooks(), ARRAY_VALID_SEPARATORS);
+    const StringSet destLookSet(destLookList.begin(), destLookList.end());
+
+    for (auto lookName : sourceLookList)
```
Comment (feldstj) [inline source/MaterialXCore/Look.cpp:208]: Should this be const auto& LookName?

File: source/MaterialXCore/Look.cpp [hunk 10]
```diff
@@ -136,4 +158,121 @@ vector<VariantAssignPtr> MaterialAssign::getActiveVariantAssigns() const
...
+    }
+
+    // If look already exists in the active look list then append if not skipping duplicates
+    //
+    string activeSourceLooks = sourceGroup->getActiveLook();
+    const StringVec& sourceActiveLookList = splitString(activeSourceLooks, ARRAY_VALID_SEPARATORS);
+    StringVec destActiveLookList = splitString(getActiveLook(), ARRAY_VALID_SEPARATORS);
+    const StringSet destActiveLookSet(destActiveLookList.begin(), destActiveLookList.end());
+
+    for (auto activeLookName : sourceActiveLookList)
```
Comment (feldstj) [inline source/MaterialXCore/Look.cpp:231]: Should this be const auto& activeLookName?

File: source/MaterialXTest/MaterialXCore/Look.cpp [hunk 11]
```diff
@@ -6,6 +6,9 @@
 #include <MaterialXTest/Catch/catch.hpp>
 
 #include <MaterialXCore/Document.h>
+#include <MaterialXFormat/XmlIo.h>
+
+#include <iostream>
```
Comment (feldstj) [inline source/MaterialXTest/MaterialXCore/Look.cpp:11]: Is this needed?

File: source/MaterialXCore/Look.cpp (hunk 8 above)
Comment (bernardkwok) [reply source/MaterialXCore/Look.cpp]: No. This is a non-const copy.

File: source/MaterialXCore/Look.cpp (hunk 9 above)
Comment (bernardkwok) [reply source/MaterialXCore/Look.cpp]: Yes. Will fix.

File: source/MaterialXCore/Look.cpp (hunk 10 above)
Comment (bernardkwok) [reply source/MaterialXCore/Look.cpp]: Yes. Will fix.

File: source/MaterialXTest/MaterialXCore/Look.cpp (hunk 11 above)
Comment (bernardkwok) [reply source/MaterialXTest/MaterialXCore/Look.cpp]: No will remove.

File: source/MaterialXCore/Look.cpp [hunk 12]
```diff
@@ -205,40 +205,61 @@ void LookGroup::append(const LookGroupPtr& sourceGroup, bool skipDuplicateLooks)
     StringVec destLookList = splitString(getLooks(), ARRAY_VALID_SEPARATORS);
     const StringSet destLookSet(destLookList.begin(), destLookList.end());
 
-    for (auto lookName : sourceLookList)
+    bool prepend = true;
+    StringVec postList;
+
+    // If inserting after a given look, then append up to that look
+    // and the append the remainder after.
```
Comment (feldstj) [inline source/MaterialXCore/Look.cpp:211]: grammar

**Review by feldstj — APPROVED** [review]

**Pull request merged**